cmd/lambda/lease_auth/auth.go

```diff
@@ -0,0 +1,185 @@
+		api.WriteAPIErrorResponse(w,
+			errors.NewInternalServer("Internal server error", err),
+		)
+		return
+	}
+
+	// Lease ID is optional.
+	// If not provided, we will attempt to find a matching
+	// lease, where PrincipalID == Cognito User ID
+	var leaseID *string
+	muxVars := mux.Vars(r)
+	if muxVars != nil {
+		leaseIDStr, ok := muxVars["leaseID"]
+		if ok {
+			leaseID = &leaseIDStr
+		}
+	}
+
+	// Generate credentials for the requested lease
+	res, err := leaseAuth(&leaseAuthInput{
+		leaseID:        leaseID,
+		requestContext: &reqCtx,
+		leaseService:   Services.LeaseService(),
+		accountService: Services.AccountService(),
+		userDetailer:   Services.UserDetailer(),
+		accountManager: Services.AccountManager(),
+		// Note that AWS has a hard 1-hour limit on roles assumed
+		// via "role-chaining" (assuming a role from an assumed-role).
+		// In our case, we're assuming a role, via our assumed lambda execution role.
+		principalSessionDuration: time.Hour,
+	})
+	if err != nil {
+		api.WriteAPIErrorResponse(w, err)
+		return
+	}
+
+	api.WriteAPIResponse(w, 201, res)
+}
+
+type leaseAuthInput struct {
+	leaseID                  *string
+	requestContext           *events.APIGatewayProxyRequestContext
+	leaseService             leaseiface.Servicer
+	accountService           accountiface.Servicer
+	userDetailer             api.UserDetailer
+	accountManager           accountmanageriface.Servicer
+	principalSessionDuration time.Duration
+}
+
+func leaseAuth(input *leaseAuthInput) (*response.LeaseAuthResponse, error) {
+	// Lookup the requesting user, via cognito
+	user := input.userDetailer.GetUser(input.requestContext)
+
+	var lease *leases.Lease
+	var err error
+	if input.leaseID != nil {
+		// Grab the lease from the DB
+		lease, err = input.leaseService.Get(*input.leaseID)
+		if err != nil {
+			return nil, err
+		}
+	} else if user.Username != "" {
+		// Lookup the active lease for the requesting user
+		leaseList, err := input.leaseService.List(&leases.Lease{
+			PrincipalID: &user.Username,
+			Status:      leases.StatusActive.StatusPtr(),
+		})
+		if err != nil {
+			return nil, err
+		}
+		leaseListSlice := []leases.Lease(*leaseList)
+		if len(leaseListSlice) >= 1 {
+			lease = &leaseListSlice[0]
+		}
+	}
+	if lease == nil {
+		return nil, errors.NewNotFound("lease for user", user.Username)
+	}
+
+	// Return auth error, if lease isn't active
+	if *lease.Status != leases.StatusActive {
+		return nil, errors.NewUnauthorized("Unable to authorize against non-active lease")
+	}
+
+	// Admin users may login to any active lease.
+	// Other users can only login to their own leases
+	isLoginAllowed := user.Role == api.AdminGroupName ||
+		user.Username == *lease.PrincipalID
+	if !isLoginAllowed {
+		return nil, errors.NewUnauthorizedf("User %s does not have access to lease %s",
+			user.Username, *lease.ID)
+	}
+
+	// Lookup the Account, so we can get the Principal Role ARN
+	acct, err := input.accountService.Get(*lease.AccountID)
```

Comment:
This isn't your work but this seems like a call we should try to remove.  Just store the principalRoleArn with the lease itself.

Reply:
> Just store the principalRoleArn with the lease itself.

@kddejong I'm not sure I'm following. Are you suggesting a change to the Lease DB table?